main.cpp

```diff
@@ -80,20 +88,62 @@ void uthread_yield();
 #endif
 
 /**
+ * Completes a thread's execution. Removes all the threads that were
+ * waiting on calling thread from the waiting list and adds to the
+ * ready list.
+ */
+void thread_complete(){
+    int ret_val = sigsetjmp(threads[current_thread_id].env,1);
+    if (ret_val == 1) return;
+
```

Comment:
Instead of calling thead_yield when a thread finishes, the thread wrapper can call this function to finish its execution.